### IMPORT PYTHON PACKAGES

In [88]:
# Import the python packages

import re
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

pd.set_option('display.max_columns', None)  
pd.set_option('display.max_rows', None)  

### READING AND UNDERSTANDING THE DATASET

In [62]:
# Read the dataset
telecom_df = pd.read_csv('telecom_churn_data.csv')

In [63]:
# Get the dataset details using head()
telecom_df.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0            9/30/2014  197.385  214.816  213.803   21.100          NaN   
1            9/30/2014   34.047  355.074  268.321   86.285        24.11   
2            9/30/2014  167.690  189.058  210.226  290.714        11.54   
3            9/30/2014  221.338  251.102  508.054  389.500        99.91   
4            9/30/2014  261.636  309.876  238.174  163.426        50.31   

   onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0          NaN         0.00          NaN           NaN           NaN   
1        78.68         7.68        18.34         15.74         99.84   
2        55.24        37.26        74.81        143.33        220.59   
3        54.39       310.98       241.71        123.31        109.01   
4       149.44        83.89        58.78         76.96         91.88   

   offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0          0.00           NaN            NaN            NaN           0.00   
1        304.76         53.76            0.0           0.00           0.00   
2        208.36        118.91            0.0           0.00           0.00   
3         71.68        113.54            0.0          54.86          44.38   
4        124.26         45.81            0.0           0.00           0.00   

   roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
0            NaN            NaN            NaN           0.00            NaN   
1           0.00            0.0           0.00           0.00           0.00   
2          38.49            0.0           0.00           0.00          70.94   
3           0.00            0.0          28.09          39.04           0.00   
4           0.00            0.0           0.00           0.00           0.00   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
0               NaN               NaN              0.00               NaN   
1             23.88             74.56              7.68             18.34   
2              7.19             28.74             13.58             14.39   
3             73.68             34.81             10.61             15.49   
4             50.31            149.44             83.89             58.78   

   loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
0               NaN               NaN              0.00               NaN   
1             11.51             75.94            291.86             53.76   
2             29.34             16.86             38.46             28.16   
3            107.43             83.21             22.46             65.46   
4             67.64             91.88            124.26             37.89   

   loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
0               NaN               NaN              0.00               NaN   
1              0.00              0.00              0.00              0.00   
2             24.11             21.79             15.61             22.24   
3              1.91              0.65              

In [64]:
# Get the shape of the dataset
telecom_df.shape

(99999, 226)

In [65]:
# Describe the dataset
telecom_df.describe(percentiles=[0.25, 0.50, 0.75, 0.90, 0.99])

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   9.999900e+04    99999.0         98981.0         98981.0   
mean    7.001207e+09      109.0             0.0             0.0   
std     6.956694e+05        0.0             0.0             0.0   
min     7.000000e+09      109.0             0.0             0.0   
25%     7.000606e+09      109.0             0.0             0.0   
50%     7.001205e+09      109.0             0.0             0.0   
75%     7.001812e+09      109.0             0.0             0.0   
90%     7.002174e+09      109.0             0.0             0.0   
99%     7.002387e+09      109.0             0.0             0.0   
max     7.002411e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8        arpu_9  \
count         98981.0  99999.000000  99999.000000  99999.000000  99999.000000   
mean              0.0    282.987358    278.536648    279.154731    261.645069   
std               0.0    328.439770    338.156291    344.474791    341.998630   
min               0.0  -2258.709000  -2014.045000   -945.808000  -1899.505000   
25%               0.0     93.411500     86.980500     84.126000     62.685000   
50%               0.0    197.704000    191.640000    192.080000    176.849000   
75%               0.0    371.060000    365.344500    369.370500    353.466500   
90%               0.0    620.075400    617.252000    618.160400    601.537000   
99%               0.0   1410.792540   1428.685880   1455.539960   1402.445660   
max               0.0  27731.088000  35145.834000  33543.624000  38805.617000   

        onnet_mou_6   onnet_mou_7   onnet_mou_8   onnet_mou_9  offnet_mou_6  \
count  96062.000000  96140.000000  94621.000000  92254.000000  96062.000000   
mean     132.395875    133.670805    133.018098    130.302327    197.935577   
std      297.207406    308.794148    308.951589    308.477668    316.851613   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        7.380000      6.660000      6.460000      5.330000     34.730000   
50%       34.310000     32.330000     32.360000     29.840000     96.310000   
75%      118.740000    115.595000    115.860000    112.130000    231.860000   
90%      344.102000    348.035000    345.030000    342.890000    477.325000   
99%     1431.760500   1478.533300   1480.334000   1445.766400   1539.535800   
max     7376.710000   8157.780000  10752.560000  10427.460000   8362.360000   

       offnet_mou_7  offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
count  96140.000000  94621.000000  92254.000000   96062.000000   96140.000000   
mean     197.045133    196.574803    190.337222       9.950013       7.149898   
std      325.862803    327.170662    319.396092      72.825411      73.447948   
min        0.000000      0.000000      0.000000       0.000000       0.000000   
25%       32.190000     31.630000     27.130000       0.000000       0.000000   
50%       91.735000     92.140000     87.290000       0.000000       0.000000   
75%      226.815000    228.260000    220.505000       0.000000       0.000000   
90%      480.431000    486.310000    471.031000      11.810000       5.060000   
99%     1549.379500   1530.398000   1521.328700     219.847000     154.483200   
max     9667.130000  14007.340000  10310.760000   13724.380000   15371.040000   

       roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  \
count   94621.000000   92254.000000   96062.000000   96140.000000   
mean        7.292981       6.343841      13.911337       9.818732   
std        68.402466      57.137537      71.443196      58.455762   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
90%         5.340000       3.630000      22.480000       9.9300

In [66]:
# Print the 'Not-Null Count' and 'Dtype' using info()
telecom_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 226 entries, mobile_number to sep_vbc_3g
dtypes: float64(179), int64(35), object(12)
memory usage: 172.4+ MB


### DATA CLEANING

In [74]:
# Checking for duplicate records.
telecom_df['mobile_number'].value_counts().sum()

99999

We can see that the number of rows is same as distinct mobile numbers, so there is no duplicate data

#### Treatment for columns with unique values

In [67]:
# Checking columns which have one unique value and dropping them as they will not affect our analysis
telecom_df.nunique()

mobile_number               99999
circle_id                       1
loc_og_t2o_mou                  1
std_og_t2o_mou                  1
loc_ic_t2o_mou                  1
last_date_of_month_6            1
last_date_of_month_7            1
last_date_of_month_8            1
last_date_of_month_9            1
arpu_6                      85681
arpu_7                      85308
arpu_8                      83615
arpu_9                      79937
onnet_mou_6                 24313
onnet_mou_7                 24336
onnet_mou_8                 24089
onnet_mou_9                 23565
offnet_mou_6                31140
offnet_mou_7                31023
offnet_mou_8                30908
offnet_mou_9                30077
roam_ic_mou_6                6512
roam_ic_mou_7                5230
roam_ic_mou_8                5315
roam_ic_mou_9                4827
roam_og_mou_6                8038
roam_og_mou_7                6639
roam_og_mou_8                6504
roam_og_mou_9                5882
loc_og_t2t_mou

In [68]:
# Drop the columns with unique values

# Code block to get the columns with unique values and drop them
telecom_df_columns = telecom_df.columns

for col in telecom_df_columns:
    if telecom_df[col].nunique() == 1:
        telecom_df.drop(col, axis=1, inplace=True)

In [69]:
telecom_df.shape

(99999, 210)

#### Treatment for null values

In [70]:
# Let's check the percentage of null values

round(100*(telecom_df.isnull().sum()/len(telecom_df.index)), 2)

mobile_number                0.00
arpu_6                       0.00
arpu_7                       0.00
arpu_8                       0.00
arpu_9                       0.00
onnet_mou_6                  3.94
onnet_mou_7                  3.86
onnet_mou_8                  5.38
onnet_mou_9                  7.75
offnet_mou_6                 3.94
offnet_mou_7                 3.86
offnet_mou_8                 5.38
offnet_mou_9                 7.75
roam_ic_mou_6                3.94
roam_ic_mou_7                3.86
roam_ic_mou_8                5.38
roam_ic_mou_9                7.75
roam_og_mou_6                3.94
roam_og_mou_7                3.86
roam_og_mou_8                5.38
roam_og_mou_9                7.75
loc_og_t2t_mou_6             3.94
loc_og_t2t_mou_7             3.86
loc_og_t2t_mou_8             5.38
loc_og_t2t_mou_9             7.75
loc_og_t2m_mou_6             3.94
loc_og_t2m_mou_7             3.86
loc_og_t2m_mou_8             5.38
loc_og_t2m_mou_9             7.75
loc_og_t2f_mou

Delete the columns where null value percentage is > 40%

In [71]:
# Code block to delete the columns having more than 40% of null values
telecom_df_columns = telecom_df.columns

for col in telecom_df_columns:
    if telecom_df[col].isnull().sum()/len(telecom_df)*100 > 40:
        telecom_df.drop(col, axis=1, inplace=True)

In [72]:
telecom_df.shape

(99999, 170)

#### Filtering High Value Customers

In [75]:
#Deriving Average recharge amount for the month of June and July.
telecom_df['Average_rech_amt_6_and_7'] = (telecom_df['total_rech_amt_6'] + telecom_df['total_rech_amt_7'])/2

In [78]:
# If the recharge amount for the month of June and July are >= 70th percentile of Average_rech_amt_6n7, such customers are HIGH VALUED CUSTOMERS
telecom_df = telecom_df[(telecom_df['Average_rech_amt_6_and_7'] >= telecom_df['Average_rech_amt_6_and_7'].quantile(0.7))]

In [81]:
telecom_df.head()

mobile_number    arpu_6    arpu_7    arpu_8   arpu_9  onnet_mou_6  \
7      7000701601  1069.180  1349.850  3171.480  500.000        57.84   
8      7001524846   378.721   492.223   137.362  166.787       413.69   
13     7002191713   492.846   205.671   593.260  322.732       501.76   
16     7000875565   430.975   299.869   187.894  206.490        50.51   
17     7000187447   690.008    18.980    25.499  257.583      1185.91   

    onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
7         54.68        52.29          NaN        453.43        567.16   
8        351.03        35.08        33.46         94.66         80.63   
13       108.39       534.24       244.81        413.31        119.28   
16        74.01        70.61        31.34        296.29        229.74   
17         9.28         7.79       558.51         61.64          0.00   

    offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
7         325.91           NaN          16.23          33.49          31.64   
8         136.48        108.71           0.00           0.00           0.00   
13        482.46        214.06          23.53         144.24          72.11   
16        162.76        224.39           0.00           2.83           0.00   
17          5.54         87.89           0.00           4.76           4.81   

    roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
7             NaN          23.74          12.59          38.06            NaN   
8            0.00           0.00           0.00           0.00           0.00   
13         136.78           7.98          35.26           1.44          12.78   
16           0.00           0.00          17.74           0.00           0.00   
17           0.00           0.00           8.46          13.34          17.98   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
7              51.39             31.38             40.28               NaN   
8             297.13            217.59             12.49             26.13   
13             49.63              6.19             36.01              6.14   
16             42.61             65.16             67.38             26.88   
17             38.99              0.00              0.00             36.41   

    loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
7             308.63            447.38            162.28               NaN   
8              80.96             70.58             50.54             34.58   
13            151.13             47.28            294.46            108.24   
16            273.29            145.99            128.28            201.49   
17             58.54              0.00              0.00              9.38   

    loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
7              62.13             55.14             53.23               NaN   
8               0.00              0.00              0.00              0.00   
13              4.54              0.00             23.51              5.29   
16              0.00              4.48             10.26              4.66   
17              0.00              0.00              0.00              0.00   

    loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  \
7                0.0               0.0              0.00               NaN   
8                0.0               0.0              7.15               0.0   
13               0.0               0.0              0.49               0.0   
16               0.0               0.0              0.00               0.0   
17               0.0               0.0              0.00               0.0   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
7         422.16        533.91        255.79           NaN              4.30   
8         378.09        288.18         63.04         60.71            116.56   
13        205.31         53.48        353.99    

In [82]:
telecom_df.shape

(30011, 171)

In [97]:
# Let's check the percentage of null values

telecom_df_missing_columns = (round(((telecom_df.isnull().sum()/len(telecom_df.index))*100),2).to_frame('Null')).sort_values('Null', ascending=False)
telecom_df_missing_columns

Null
std_og_t2f_mou_9          5.68
roam_ic_mou_9             5.68
loc_og_mou_9              5.68
loc_ic_t2t_mou_9          5.68
loc_og_t2f_mou_9          5.68
std_ic_t2m_mou_9          5.68
std_ic_mou_9              5.68
loc_og_t2m_mou_9          5.68
std_og_t2t_mou_9          5.68
spl_ic_mou_9              5.68
loc_og_t2t_mou_9          5.68
loc_ic_t2m_mou_9          5.68
std_og_t2m_mou_9          5.68
roam_og_mou_9             5.68
og_others_9               5.68
loc_og_t2c_mou_9          5.68
std_ic_t2t_mou_9          5.68
ic_others_9               5.68
loc_ic_t2f_mou_9          5.68
offnet_mou_9              5.68
std_og_mou_9              5.68
spl_og_mou_9              5.68
isd_ic_mou_9              5.68
loc_ic_mou_9              5.68
isd_og_mou_9              5.68
onnet_mou_9               5.68
std_ic_t2f_mou_9          5.68
isd_og_mou_8              3.13
loc_og_mou_8              3.13
og_others_8               3.13
std_og_t2t_mou_8          3.13
std_og_mou_8              3.13
std_ic_mou_8              3.13
std_og_t2m_mou_8          3.13
std_og_t2f_mou_8          3.13
std_ic_t2m_mou_8          3.13
loc_og_t2f_mou_8          3.13
std_ic_t2f_mou_8          3.13
loc_og_t2t_mou_8          3.13
std_ic_t2t_mou_8          3.13
isd_ic_mou_8              3.13
roam_ic_mou_8             3.13
loc_ic_t2f_mou_8          3.13
onnet_mou_8               3.13
roam_og_mou_8             3.13
loc_og_t2c_mou_8          3.13
offnet_mou_8              3.13
loc_ic_t2m_mou_8          3.13
loc_ic_mou_8              3.13
loc_og_t2m_mou_8          3.13
ic_others_8               3.13
spl_ic_mou_8              3.13
spl_og_mou_8              3.13
loc_ic_t2t_mou_8          3.13
date_of_last_rech_9       2.89
date_of_last_rech_8       1.98
og_others_6               1.05
loc_ic_t2t_mou_6          1.05
std_ic_t2m_mou_6          1.05
loc_ic_t2f_mou_6          1.05
std_ic_t2f_mou_6          1.05
loc_ic_t2m_mou_6          1.05
loc_ic_mou_6              1.05
std_og_t2m_mou_6          1.05
spl_og_mou_6              1.05
loc_og_t2c_mou_6          1.05
onnet_mou_6               1.05
ic_others_6               1.05
offnet_mou_6              1.05
roam_ic_mou_6             1.05
roam_og_mou_6             1.05
loc_og_t2t_mou_6          1.05
isd_ic_mou_6              1.05
loc_og_t2m_mou_6          1.05
isd_og_mou_6              1.05
loc_og_t2f_mou_6          1.05
loc_og_mou_6              1.05
spl_ic_mou_6              1.05
std_og_t2t_mou_6          1.05
std_og_t2f_mou_6          1.05
std_og_mou_6              1.05
std_ic_mou_6              1.05
std_ic_t2t_mou_6          1.05
spl_ic_mou_7              1.01
std_ic_t2m_mou_7          1.01
std_ic_t2t_mou_7          1.01
isd_ic_mou_7              1.01
std_ic_mou_7              1.01
std_ic_t2f_mou_7          1.01
ic_others_7               1.01
spl_og_mou_7              1.01
loc_ic_mou_7              1.01
std_og_t2f_mou_7          1.01
roam_ic_mou_7             1.01
roam_og_mou_7             1.01
loc_og_t2t_mou_7          1.01
loc_og_t2m_mou_7          1.01
loc_og_t2f_mou_7          1.01
loc_og_t2c_mou_7          1.01
loc_og_mou_7              1.01
onnet_mou_7               1.01
std_og_t2m_mou_7          1.01
std_og_t2t_mou_7          1.01
std_og_mou_7              1.01
isd_og_mou_7              1.01
og_others_7               1.01
loc_ic_t2t_mou_7          1.01
loc_ic_t2m_mou_7          1.01
loc_ic_t2f_mou_7          1.01
offnet_mou_7              1.01
date_of_last_rech_7       0.38
date_of_last_rech_6       0.21
monthly_2g_8              0.00
vol_2g_mb_8               0.00
vol_2g_mb_9               0.00
vol_3g_mb_6               0.00
monthly_2g_9              0.00
monthly_2g_6              0.00
monthly_2g_7              0.00
vol_3g_mb_7               0.00
vol_3g_mb_9               0.00
vol_3g_mb_8               0.00
sachet_2g_7               0.00
sachet_2g_6               0.00
aon                       0.00
sachet_2g_8               0.00
sachet_2g_9               0.00
monthly_3g_6              0.00
monthly_3g_7              0

In [94]:
# Let's handle the missing values month wise

#### September

In [112]:
sep_missing_cols = (telecom_df_missing_columns[telecom_df_missing_columns['Null'] == 5.68].index).to_list()
sep_missing_cols

['std_og_t2f_mou_9',
 'roam_ic_mou_9',
 'loc_og_mou_9',
 'loc_ic_t2t_mou_9',
 'loc_og_t2f_mou_9',
 'std_ic_t2m_mou_9',
 'std_ic_mou_9',
 'loc_og_t2m_mou_9',
 'std_og_t2t_mou_9',
 'spl_ic_mou_9',
 'loc_og_t2t_mou_9',
 'loc_ic_t2m_mou_9',
 'std_og_t2m_mou_9',
 'roam_og_mou_9',
 'og_others_9',
 'loc_og_t2c_mou_9',
 'std_ic_t2t_mou_9',
 'ic_others_9',
 'loc_ic_t2f_mou_9',
 'offnet_mou_9',
 'std_og_mou_9',
 'spl_og_mou_9',
 'isd_ic_mou_9',
 'loc_ic_mou_9',
 'isd_og_mou_9',
 'onnet_mou_9',
 'std_ic_t2f_mou_9']

In [126]:
# Creating a dataframe for MOU for September, for which the the MOU values are null
telecom_df_null_mou_9 = telecom_df[(telecom_df['std_og_t2f_mou_9'].isnull()) & 
                                    (telecom_df['roam_ic_mou_9'].isnull()) & 
                                    (telecom_df['loc_og_mou_9'].isnull()) & 
                                    (telecom_df['loc_ic_t2t_mou_9'].isnull()) & 
                                    (telecom_df['loc_og_t2f_mou_9'].isnull()) & 
                                    (telecom_df['std_ic_t2m_mou_9'].isnull()) & 
                                    (telecom_df['std_ic_mou_9'].isnull()) & 
                                    (telecom_df['loc_og_t2m_mou_9'].isnull()) & 
                                    (telecom_df['std_og_t2t_mou_9'].isnull()) & 
                                    (telecom_df['spl_ic_mou_9'].isnull()) & 
                                    (telecom_df['loc_og_t2t_mou_9'].isnull()) & 
                                    (telecom_df['loc_ic_t2m_mou_9'].isnull()) & 
                                    (telecom_df['std_og_t2m_mou_9'].isnull()) & 
                                    (telecom_df['roam_og_mou_9'].isnull()) & 
                                    (telecom_df['og_others_9'].isnull()) & 
                                    (telecom_df['loc_og_t2c_mou_9'].isnull()) & 
                                    (telecom_df['std_ic_t2t_mou_9'].isnull()) & 
                                    (telecom_df['ic_others_9'].isnull()) & 
                                    (telecom_df['loc_ic_t2f_mou_9'].isnull()) & 
                                    (telecom_df['offnet_mou_9'].isnull()) &
                                    (telecom_df['std_og_mou_9'].isnull()) & 
                                    (telecom_df['spl_og_mou_9'].isnull()) & 
                                    (telecom_df['isd_ic_mou_9'].isnull()) & 
                                    (telecom_df['loc_ic_mou_9'].isnull()) & 
                                    (telecom_df['isd_og_mou_9'].isnull()) & 
                                    (telecom_df['onnet_mou_9'].isnull()) & 
                                    (telecom_df['std_ic_t2f_mou_9'].isnull())
                                     ]

telecom_df_null_mou_9.head()

mobile_number    arpu_6    arpu_7    arpu_8  arpu_9  onnet_mou_6  \
7       7000701601  1069.180  1349.850  3171.480   500.0        57.84   
97      7000589828   374.863   294.023   183.043     0.0       433.59   
111     7001300706   596.301   146.073     0.000     0.0        55.19   
143     7000106299   695.609    39.981     0.000     0.0      1325.91   
188     7000340381   734.641   183.668     0.000     0.0         4.38   

     onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
7          54.68        52.29          NaN        453.43        567.16   
97        415.66       221.06          NaN         74.54         43.66   
111         3.26          NaN          NaN         45.51         12.34   
143        28.61          NaN          NaN         13.91          1.89   
188         0.98          NaN          NaN        105.16         39.39   

     offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
7          325.91           NaN          16.23          33.49          31.64   
97          31.86           NaN           0.00           0.00           6.16   
111           NaN           NaN           0.00           0.00            NaN   
143           NaN           NaN           0.00           8.94            NaN   
188           NaN           NaN           0.00           0.00            NaN   

     roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
7              NaN          23.74          12.59          38.06   
97             NaN           0.00           0.00          23.91   
111            NaN           0.00           0.00            NaN   
143            NaN           0.00           8.53            NaN   
188            NaN           0.00           0.00            NaN   

     roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
7              NaN             51.39             31.38             40.28   
97             NaN              2.83             16.19              9.73   
111            NaN             55.19              3.26               NaN   
143            NaN             18.89              6.83               NaN   
188            NaN              4.38              0.98               NaN   

     loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
7                 NaN            308.63            447.38            162.28   
97                NaN             16.99             23.14             17.79   
111               NaN             43.83             12.34               NaN   
143               NaN              8.58              1.56               NaN   
188               NaN             99.81             38.98               NaN   

     loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
7                 NaN             62.13             55.14             53.23   
97                NaN              3.54              1.46              1.83   
111               NaN              0.00              0.00               NaN   
143               NaN              0.00              0.00               NaN   
188               NaN              5.34              0.41               NaN   

     loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
7                 NaN              0.00               0.0               0.0   
97                NaN              0.40               0.0               0.0   
111               NaN              0.00               0.0               NaN   
143               NaN              2.09               0.0               NaN   
188               NaN              0.00               0.0               NaN   

     loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  \
7                 NaN        422.16        533.91        255.79           NaN   
97                NaN         23.38         40.81         29.36           NaN   
111               NaN         99.03         15.61           NaN           NaN   
143               NaN         2

In [127]:
telecom_df_null_mou_9.shape

(1704, 171)

In [128]:
# Droping the records for which MOU for September are null
telecom_df = telecom_df.drop(telecom_df_null_mou_9.index)

In [129]:
telecom_df.shape

(28307, 171)

In [130]:
# Let's check the percentage of null values again

telecom_df_missing_columns = (round(((telecom_df.isnull().sum()/len(telecom_df.index))*100),2).to_frame('Null')).sort_values('Null', ascending=False)
telecom_df_missing_columns

Null
date_of_last_rech_8       1.08
date_of_last_rech_9       1.04
std_ic_t2f_mou_8          0.55
std_ic_t2t_mou_8          0.55
std_ic_mou_8              0.55
loc_og_t2t_mou_8          0.55
std_og_t2f_mou_8          0.55
loc_ic_t2f_mou_8          0.55
loc_og_t2m_mou_8          0.55
std_og_t2m_mou_8          0.55
loc_ic_mou_8              0.55
loc_og_t2f_mou_8          0.55
og_others_8               0.55
std_ic_t2m_mou_8          0.55
loc_og_t2c_mou_8          0.55
std_og_t2t_mou_8          0.55
roam_og_mou_8             0.55
loc_og_mou_8              0.55
loc_ic_t2t_mou_8          0.55
roam_ic_mou_8             0.55
loc_ic_t2m_mou_8          0.55
ic_others_8               0.55
spl_ic_mou_8              0.55
offnet_mou_8              0.55
isd_og_mou_8              0.55
std_og_mou_8              0.55
onnet_mou_8               0.55
isd_ic_mou_8              0.55
spl_og_mou_8              0.55
loc_ic_t2f_mou_6          0.50
std_og_t2t_mou_6          0.50
std_ic_t2t_mou_6          0.50
og_others_6               0.50
spl_og_mou_6              0.50
std_og_t2m_mou_6          0.50
loc_ic_t2m_mou_6          0.50
std_og_t2f_mou_6          0.50
isd_og_mou_6              0.50
std_og_mou_6              0.50
loc_ic_mou_6              0.50
loc_og_t2c_mou_6          0.50
std_ic_t2m_mou_6          0.50
loc_og_t2m_mou_6          0.50
ic_others_6               0.50
onnet_mou_6               0.50
offnet_mou_6              0.50
isd_ic_mou_6              0.50
roam_ic_mou_6             0.50
spl_ic_mou_6              0.50
loc_og_t2t_mou_6          0.50
roam_og_mou_6             0.50
std_ic_mou_6              0.50
loc_og_t2f_mou_6          0.50
std_ic_t2f_mou_6          0.50
loc_og_mou_6              0.50
loc_ic_t2t_mou_6          0.50
std_ic_mou_7              0.23
loc_ic_t2f_mou_7          0.23
std_ic_t2t_mou_7          0.23
loc_ic_mou_7              0.23
spl_ic_mou_7              0.23
loc_ic_t2m_mou_7          0.23
std_ic_t2f_mou_7          0.23
isd_ic_mou_7              0.23
std_ic_t2m_mou_7          0.23
loc_ic_t2t_mou_7          0.23
ic_others_7               0.23
spl_og_mou_7              0.23
offnet_mou_7              0.23
loc_og_mou_7              0.23
std_og_t2m_mou_7          0.23
std_og_t2t_mou_7          0.23
loc_og_t2m_mou_7          0.23
std_og_t2f_mou_7          0.23
std_og_mou_7              0.23
onnet_mou_7               0.23
roam_og_mou_7             0.23
isd_og_mou_7              0.23
loc_og_t2c_mou_7          0.23
og_others_7               0.23
loc_og_t2f_mou_7          0.23
loc_og_t2t_mou_7          0.23
roam_ic_mou_7             0.23
date_of_last_rech_7       0.20
date_of_last_rech_6       0.17
sachet_3g_9               0.00
sachet_3g_8               0.00
last_day_rch_amt_6        0.00
last_day_rch_amt_7        0.00
monthly_2g_8              0.00
aon                       0.00
aug_vbc_3g                0.00
max_rech_amt_9            0.00
jul_vbc_3g                0.00
jun_vbc_3g                0.00
sep_vbc_3g                0.00
max_rech_amt_8            0.00
max_rech_amt_7            0.00
max_rech_amt_6            0.00
total_rech_amt_9          0.00
sachet_3g_7               0.00
last_day_rch_amt_9        0.00
last_day_rch_amt_8        0.00
sachet_2g_8               0.00
monthly_2g_6              0.00
vol_3g_mb_9               0.00
vol_3g_mb_8               0.00
monthly_2g_9              0.00
total_rech_amt_7          0.00
vol_3g_mb_7               0.00
sachet_2g_6               0.00
sachet_2g_7               0.00
vol_3g_mb_6               0.00
vol_2g_mb_9               0.00
sachet_3g_6               0.00
vol_2g_mb_8               0.00
sachet_2g_9               0.00
vol_2g_mb_7               0.00
monthly_3g_6              0.00
monthly_3g_7              0.00
vol_2g_mb_6               0.00
monthly_3g_8              0.00
monthly_2g_7              0.00
monthly_3g_9              0.00
total_rech_amt_8          0.00
mobile_number             0.00
total_rech_amt_6          0.00
total_og_mou_7            0.00
og_others_9               0

#### August

In [133]:
aug_missing_cols = (telecom_df_missing_columns[telecom_df_missing_columns['Null'] == 0.55].index).to_list()
aug_missing_cols

['std_ic_t2f_mou_8',
 'std_ic_t2t_mou_8',
 'std_ic_mou_8',
 'loc_og_t2t_mou_8',
 'std_og_t2f_mou_8',
 'loc_ic_t2f_mou_8',
 'loc_og_t2m_mou_8',
 'std_og_t2m_mou_8',
 'loc_ic_mou_8',
 'loc_og_t2f_mou_8',
 'og_others_8',
 'std_ic_t2m_mou_8',
 'loc_og_t2c_mou_8',
 'std_og_t2t_mou_8',
 'roam_og_mou_8',
 'loc_og_mou_8',
 'loc_ic_t2t_mou_8',
 'roam_ic_mou_8',
 'loc_ic_t2m_mou_8',
 'ic_others_8',
 'spl_ic_mou_8',
 'offnet_mou_8',
 'isd_og_mou_8',
 'std_og_mou_8',
 'onnet_mou_8',
 'isd_ic_mou_8',
 'spl_og_mou_8']

In [134]:
# Creating a dataframe for MOU for August, for which the the MOU values are null
telecom_df_null_mou_8 = telecom_df[(telecom_df['std_ic_t2f_mou_8'].isnull()) & 
                                    (telecom_df['std_ic_t2t_mou_8'].isnull()) & 
                                    (telecom_df['std_ic_mou_8'].isnull()) & 
                                    (telecom_df['loc_og_t2t_mou_8'].isnull()) & 
                                    (telecom_df['std_og_t2f_mou_8'].isnull()) & 
                                    (telecom_df['loc_ic_t2f_mou_8'].isnull()) & 
                                    (telecom_df['loc_og_t2m_mou_8'].isnull()) & 
                                    (telecom_df['std_og_t2m_mou_8'].isnull()) & 
                                    (telecom_df['loc_ic_mou_8'].isnull()) & 
                                    (telecom_df['loc_og_t2f_mou_8'].isnull()) & 
                                    (telecom_df['og_others_8'].isnull()) & 
                                    (telecom_df['std_ic_t2m_mou_8'].isnull()) & 
                                    (telecom_df['loc_og_t2c_mou_8'].isnull()) & 
                                    (telecom_df['std_og_t2t_mou_8'].isnull()) & 
                                    (telecom_df['roam_og_mou_8'].isnull()) & 
                                    (telecom_df['loc_og_mou_8'].isnull()) & 
                                    (telecom_df['loc_ic_t2t_mou_8'].isnull()) & 
                                    (telecom_df['roam_ic_mou_8'].isnull()) & 
                                    (telecom_df['loc_ic_t2m_mou_8'].isnull()) & 
                                    (telecom_df['ic_others_8'].isnull()) &
                                    (telecom_df['spl_ic_mou_8'].isnull()) & 
                                    (telecom_df['offnet_mou_8'].isnull()) & 
                                    (telecom_df['isd_og_mou_8'].isnull()) & 
                                    (telecom_df['std_og_mou_8'].isnull()) & 
                                    (telecom_df['onnet_mou_8'].isnull()) & 
                                    (telecom_df['isd_ic_mou_8'].isnull()) & 
                                    (telecom_df['spl_og_mou_8'].isnull())]

telecom_df_null_mou_8.head()

mobile_number   arpu_6   arpu_7  arpu_8   arpu_9  onnet_mou_6  \
375      7002252754  580.477  111.878     0.0  378.881       249.43   
578      7000248548  569.612  237.289     0.0    4.440       718.01   
788      7000636808  532.742  546.756     0.0  269.274      1173.39   
1802     7000516213  810.455    0.000     0.0    0.000        91.33   
4837     7002192662  649.150  149.572     0.0    0.250      1354.24   

      onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
375         39.64          NaN       245.06         62.24         37.24   
578        212.73          NaN         0.00        487.06        139.71   
788        891.83          NaN       149.34         61.59        137.14   
1802          NaN          NaN         0.00       1371.04           NaN   
4837        85.13          NaN         0.43         50.63         37.13   

      offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
375            NaN        144.53          25.49          19.43            NaN   
578            NaN          1.26           0.00           2.01            NaN   
788            NaN        428.36           0.00           1.48            NaN   
1802           NaN          0.00           1.21            NaN            NaN   
4837           NaN          0.00           0.00          12.84            NaN   

      roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
375            0.00         312.59          78.58            NaN   
578            6.43           0.00           6.30            NaN   
788            0.00           0.00          14.43            NaN   
1802           0.00          11.23            NaN            NaN   
4837           1.25           0.00          44.78            NaN   

      roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
375            0.00              0.00              0.00               NaN   
578            1.26             11.28             27.89               NaN   
788            0.00             31.06             27.49               NaN   
1802           3.91             17.86               NaN               NaN   
4837           0.43              6.71              1.35               NaN   

      loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
375              11.54              0.00              0.00               NaN   
578               0.00             42.24             46.94               NaN   
788               7.39             34.66             60.86               NaN   
1802              0.00             84.51               NaN               NaN   
4837              0.00             15.18             15.76               NaN   

      loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
375              25.31               0.0               0.0               NaN   
578               0.00               0.0               0.0               NaN   
788              34.23               0.0               0.0               NaN   
1802              0.00               0.0               NaN               NaN   
4837              0.00               0.0               0.0               NaN   

      loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
375                0.0              0.00               0.0               NaN   
578                0.0              2.33               0.0               NaN   
788                0.0              0.00               0.0               NaN   
1802               0.0             10.29               NaN               NaN   
4837               0.0              0.00               0.0               NaN   

      loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
375               0.41          0.00          0.00           NaN   
578               0.00         53.53         74.84           NaN   
788               5.58         65.73         88.36           NaN   
1802              0.00        102.38           NaN    

In [135]:
telecom_df_null_mou_8.shape

(157, 171)

In [136]:
# Droping the records for which MOU for August are null
telecom_df = telecom_df.drop(telecom_df_null_mou_8.index)

In [137]:
# Let's check the percentage of null values again

telecom_df_missing_columns = (round(((telecom_df.isnull().sum()/len(telecom_df.index))*100),2).to_frame('Null')).sort_values('Null', ascending=False)
telecom_df_missing_columns

Null
date_of_last_rech_9       1.02
date_of_last_rech_8       1.01
loc_ic_mou_6              0.44
og_others_6               0.44
loc_og_t2m_mou_6          0.44
loc_ic_t2t_mou_6          0.44
loc_og_t2f_mou_6          0.44
loc_og_t2c_mou_6          0.44
spl_og_mou_6              0.44
loc_ic_t2f_mou_6          0.44
loc_og_mou_6              0.44
isd_og_mou_6              0.44
spl_ic_mou_6              0.44
std_og_t2t_mou_6          0.44
std_og_mou_6              0.44
std_og_t2f_mou_6          0.44
loc_og_t2t_mou_6          0.44
loc_ic_t2m_mou_6          0.44
std_og_t2m_mou_6          0.44
std_ic_t2m_mou_6          0.44
offnet_mou_6              0.44
std_ic_t2f_mou_6          0.44
std_ic_mou_6              0.44
onnet_mou_6               0.44
roam_ic_mou_6             0.44
ic_others_6               0.44
isd_ic_mou_6              0.44
std_ic_t2t_mou_6          0.44
roam_og_mou_6             0.44
date_of_last_rech_7       0.20
date_of_last_rech_6       0.17
std_ic_mou_7              0.16
std_ic_t2f_mou_7          0.16
isd_ic_mou_7              0.16
std_og_t2f_mou_7          0.16
isd_og_mou_7              0.16
loc_ic_mou_7              0.16
spl_og_mou_7              0.16
loc_ic_t2f_mou_7          0.16
ic_others_7               0.16
std_ic_t2m_mou_7          0.16
std_ic_t2t_mou_7          0.16
loc_ic_t2t_mou_7          0.16
std_og_t2m_mou_7          0.16
loc_ic_t2m_mou_7          0.16
og_others_7               0.16
std_og_mou_7              0.16
loc_og_t2t_mou_7          0.16
loc_og_mou_7              0.16
onnet_mou_7               0.16
offnet_mou_7              0.16
roam_ic_mou_7             0.16
roam_og_mou_7             0.16
loc_og_t2m_mou_7          0.16
loc_og_t2f_mou_7          0.16
loc_og_t2c_mou_7          0.16
spl_ic_mou_7              0.16
std_og_t2t_mou_7          0.16
monthly_2g_7              0.00
total_rech_amt_6          0.00
vol_2g_mb_8               0.00
total_rech_num_8          0.00
total_rech_num_7          0.00
sachet_3g_6               0.00
sachet_3g_7               0.00
total_rech_num_6          0.00
ic_others_9               0.00
vol_2g_mb_9               0.00
sachet_3g_8               0.00
ic_others_8               0.00
sachet_3g_9               0.00
vol_3g_mb_9               0.00
aon                       0.00
aug_vbc_3g                0.00
isd_ic_mou_9              0.00
vol_3g_mb_8               0.00
jul_vbc_3g                0.00
isd_ic_mou_8              0.00
jun_vbc_3g                0.00
sep_vbc_3g                0.00
spl_ic_mou_9              0.00
spl_ic_mou_8              0.00
total_rech_num_9          0.00
total_rech_amt_7          0.00
vol_2g_mb_6               0.00
sachet_2g_8               0.00
last_day_rch_amt_9        0.00
last_day_rch_amt_8        0.00
monthly_2g_8              0.00
monthly_2g_9              0.00
last_day_rch_amt_7        0.00
last_day_rch_amt_6        0.00
vol_2g_mb_7               0.00
sachet_2g_6               0.00
sachet_2g_7               0.00
monthly_2g_6              0.00
sachet_2g_9               0.00
monthly_3g_9              0.00
max_rech_amt_9            0.00
max_rech_amt_8            0.00
monthly_3g_6              0.00
max_rech_amt_7            0.00
max_rech_amt_6            0.00
monthly_3g_7              0.00
vol_3g_mb_6               0.00
total_rech_amt_9          0.00
total_rech_amt_8          0.00
monthly_3g_8              0.00
vol_3g_mb_7               0.00
mobile_number             0.00
total_ic_mou_9            0.00
std_og_mou_9              0.00
std_og_t2f_mou_9          0.00
std_og_t2f_mou_8          0.00
std_og_t2m_mou_9          0.00
std_og_t2m_mou_8          0.00
std_og_t2t_mou_9          0.00
std_og_t2t_mou_8          0.00
loc_og_mou_9              0.00
loc_og_mou_8              0.00
loc_og_t2c_mou_9          0.00
loc_og_t2c_mou_8          0.00
loc_og_t2f_mou_9          0.00
loc_og_t2f_mou_8          0.00
loc_og_t2m_mou_9          0.00
loc_og_t2m_mou_8          0.00
loc_og_t2t_mou_9          0.00
loc_og_t2t_mou_8          0.00
roam_og_mou_9             0

#### July

In [141]:
july_missing_cols = (telecom_df_missing_columns[telecom_df_missing_columns['Null'] == 0.16].index).to_list()
july_missing_cols

['std_ic_mou_7',
 'std_ic_t2f_mou_7',
 'isd_ic_mou_7',
 'std_og_t2f_mou_7',
 'isd_og_mou_7',
 'loc_ic_mou_7',
 'spl_og_mou_7',
 'loc_ic_t2f_mou_7',
 'ic_others_7',
 'std_ic_t2m_mou_7',
 'std_ic_t2t_mou_7',
 'loc_ic_t2t_mou_7',
 'std_og_t2m_mou_7',
 'loc_ic_t2m_mou_7',
 'og_others_7',
 'std_og_mou_7',
 'loc_og_t2t_mou_7',
 'loc_og_mou_7',
 'onnet_mou_7',
 'offnet_mou_7',
 'roam_ic_mou_7',
 'roam_og_mou_7',
 'loc_og_t2m_mou_7',
 'loc_og_t2f_mou_7',
 'loc_og_t2c_mou_7',
 'spl_ic_mou_7',
 'std_og_t2t_mou_7']

In [142]:
# Creating a dataframe for MOU for August, for which the the MOU values are null
telecom_df_null_mou_7 = telecom_df[(telecom_df['std_ic_mou_7'].isnull()) & 
                                    (telecom_df['std_ic_t2f_mou_7'].isnull()) & 
                                    (telecom_df['isd_ic_mou_7'].isnull()) & 
                                    (telecom_df['std_og_t2f_mou_7'].isnull()) & 
                                    (telecom_df['isd_og_mou_7'].isnull()) & 
                                    (telecom_df['loc_ic_mou_7'].isnull()) & 
                                    (telecom_df['spl_og_mou_7'].isnull()) & 
                                    (telecom_df['loc_ic_t2f_mou_7'].isnull()) & 
                                    (telecom_df['ic_others_7'].isnull()) & 
                                    (telecom_df['std_ic_t2m_mou_7'].isnull()) & 
                                    (telecom_df['std_ic_t2t_mou_7'].isnull()) & 
                                    (telecom_df['loc_ic_t2t_mou_7'].isnull()) & 
                                    (telecom_df['std_og_t2m_mou_7'].isnull()) & 
                                    (telecom_df['loc_ic_t2m_mou_7'].isnull()) & 
                                    (telecom_df['og_others_7'].isnull()) & 
                                    (telecom_df['std_og_mou_7'].isnull()) & 
                                    (telecom_df['loc_og_t2t_mou_7'].isnull()) & 
                                    (telecom_df['loc_og_mou_7'].isnull()) & 
                                    (telecom_df['onnet_mou_7'].isnull()) & 
                                    (telecom_df['offnet_mou_7'].isnull()) &
                                    (telecom_df['roam_ic_mou_7'].isnull()) & 
                                    (telecom_df['roam_og_mou_7'].isnull()) & 
                                    (telecom_df['loc_og_t2m_mou_7'].isnull()) & 
                                    (telecom_df['loc_og_t2f_mou_7'].isnull()) & 
                                    (telecom_df['loc_og_t2c_mou_7'].isnull()) & 
                                    (telecom_df['spl_ic_mou_7'].isnull()) & 
                                    (telecom_df['std_og_t2t_mou_7'].isnull())]

telecom_df_null_mou_7.head()

mobile_number    arpu_6   arpu_7    arpu_8    arpu_9  onnet_mou_6  \
5616      7001238202   760.815  531.088   992.818  1144.676       324.91   
9021      7000563525   266.914  722.042  1078.905   247.032          NaN   
9451      7001477649  1129.566    0.000   128.252   802.648        11.89   
9955      7001658068   925.028  189.000   789.761   445.707        46.39   
10274     7000019836   428.933  693.332   351.477   622.228          NaN   

       onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
5616           NaN       386.13      1180.29        350.29           NaN   
9021           NaN         0.00         0.00           NaN           NaN   
9451           NaN         1.46        33.89        259.18           NaN   
9955           NaN        43.39        56.61        333.78           NaN   
10274          NaN        14.93        31.24           NaN           NaN   

       offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
5616         399.64        887.76         463.63            NaN   
9021           0.00          0.00            NaN            NaN   
9451          26.21        241.18           9.98            NaN   
9955         196.53        144.73           0.00            NaN   
10274        526.11       1006.01            NaN            NaN   

       roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  \
5616          221.46            0.0         505.71            NaN   
9021            0.00            0.0            NaN            NaN   
9451            1.73            0.0           5.66            NaN   
9955            0.00            0.0           0.00            NaN   
10274           0.00            0.0            NaN            NaN   

       roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
5616          175.93            0.0            145.91               NaN   
9021            0.00            0.0               NaN               NaN   
9451            2.46            0.0              6.73               NaN   
9955            0.00            0.0             46.39               NaN   
10274           0.00            0.0               NaN               NaN   

       loc_og_t2t_mou_8  loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
5616             243.43           1108.38              0.85               NaN   
9021               0.00              0.00               NaN               NaN   
9451               1.46             20.84            171.46               NaN   
9955              43.39             56.61            227.91               NaN   
10274              0.63              5.83               NaN               NaN   

       loc_og_t2m_mou_8  loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
5616             184.78            300.19              1.13               NaN   
9021               0.00              0.00               NaN               NaN   
9451              20.54            148.88              0.00               NaN   
9955             163.68            121.54            104.69               NaN   
10274            250.56            536.78               NaN               NaN   

       loc_og_t2f_mou_8  loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  \
5616               7.94             67.11               0.0               NaN   
9021               0.00              0.00               NaN               NaN   
9451               0.00              0.00               0.0               NaN   
9955              28.96             21.04               0.0               NaN   
10274              0.00              0.00               NaN               NaN   

       loc_og_t2c_mou_8  loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  \
5616              12.51             18.89        147.89           NaN   
9021               0.00              0.00           NaN           NaN   
9451               0.00              0.00        178.19           NaN   
9955               0.00              0.00        379.01           NaN  

In [143]:
telecom_df_null_mou_7.shape

(44, 171)

In [144]:
# Droping the records for which MOU for July are null
telecom_df = telecom_df.drop(telecom_df_null_mou_7.index)

In [145]:
telecom_df.shape

(28106, 171)

In [146]:
# Let's check the percentage of null values again

telecom_df_missing_columns = (round(((telecom_df.isnull().sum()/len(telecom_df.index))*100),2).to_frame('Null')).sort_values('Null', ascending=False)
telecom_df_missing_columns

Null
date_of_last_rech_9       1.02
date_of_last_rech_8       1.01
loc_ic_mou_6              0.41
spl_og_mou_6              0.41
loc_og_t2c_mou_6          0.41
loc_og_mou_6              0.41
std_og_t2t_mou_6          0.41
std_og_t2m_mou_6          0.41
std_og_t2f_mou_6          0.41
std_og_mou_6              0.41
isd_og_mou_6              0.41
og_others_6               0.41
loc_og_t2m_mou_6          0.41
ic_others_6               0.41
loc_ic_t2t_mou_6          0.41
isd_ic_mou_6              0.41
loc_ic_t2m_mou_6          0.41
spl_ic_mou_6              0.41
loc_ic_t2f_mou_6          0.41
std_ic_t2t_mou_6          0.41
std_ic_mou_6              0.41
std_ic_t2m_mou_6          0.41
loc_og_t2f_mou_6          0.41
std_ic_t2f_mou_6          0.41
roam_og_mou_6             0.41
onnet_mou_6               0.41
roam_ic_mou_6             0.41
offnet_mou_6              0.41
loc_og_t2t_mou_6          0.41
date_of_last_rech_7       0.20
date_of_last_rech_6       0.17
isd_ic_mou_9              0.00
sachet_3g_9               0.00
isd_ic_mou_7              0.00
isd_ic_mou_8              0.00
sachet_3g_8               0.00
aon                       0.00
ic_others_7               0.00
ic_others_8               0.00
ic_others_9               0.00
total_rech_num_6          0.00
sachet_3g_7               0.00
total_rech_num_7          0.00
spl_ic_mou_9              0.00
spl_ic_mou_8              0.00
total_rech_num_9          0.00
std_ic_mou_7              0.00
std_ic_t2f_mou_7          0.00
sep_vbc_3g                0.00
std_ic_t2f_mou_8          0.00
std_ic_t2f_mou_9          0.00
jun_vbc_3g                0.00
jul_vbc_3g                0.00
std_ic_mou_8              0.00
spl_ic_mou_7              0.00
std_ic_mou_9              0.00
total_ic_mou_6            0.00
total_ic_mou_7            0.00
total_ic_mou_8            0.00
total_ic_mou_9            0.00
aug_vbc_3g                0.00
total_rech_num_8          0.00
monthly_2g_9              0.00
sachet_3g_6               0.00
sachet_2g_9               0.00
monthly_2g_8              0.00
monthly_2g_7              0.00
monthly_2g_6              0.00
sachet_2g_7               0.00
vol_3g_mb_9               0.00
vol_3g_mb_8               0.00
vol_3g_mb_7               0.00
vol_3g_mb_6               0.00
vol_2g_mb_9               0.00
vol_2g_mb_8               0.00
sachet_2g_8               0.00
vol_2g_mb_7               0.00
vol_2g_mb_6               0.00
last_day_rch_amt_9        0.00
last_day_rch_amt_8        0.00
sachet_2g_6               0.00
last_day_rch_amt_7        0.00
last_day_rch_amt_6        0.00
monthly_3g_6              0.00
monthly_3g_7              0.00
std_ic_t2m_mou_9          0.00
monthly_3g_8              0.00
max_rech_amt_9            0.00
max_rech_amt_8            0.00
monthly_3g_9              0.00
max_rech_amt_7            0.00
max_rech_amt_6            0.00
total_rech_amt_9          0.00
total_rech_amt_8          0.00
total_rech_amt_7          0.00
total_rech_amt_6          0.00
mobile_number             0.00
std_ic_t2m_mou_8          0.00
loc_og_t2t_mou_9          0.00
loc_og_t2m_mou_8          0.00
loc_og_t2m_mou_9          0.00
loc_og_t2f_mou_7          0.00
loc_og_t2f_mou_8          0.00
loc_og_t2f_mou_9          0.00
loc_og_t2c_mou_7          0.00
loc_og_t2c_mou_8          0.00
loc_og_t2c_mou_9          0.00
loc_og_mou_7              0.00
loc_og_mou_8              0.00
loc_og_mou_9              0.00
std_og_t2t_mou_7          0.00
std_og_t2t_mou_8          0.00
std_og_t2t_mou_9          0.00
std_og_t2m_mou_7          0.00
loc_og_t2m_mou_7          0.00
loc_og_t2t_mou_8          0.00
std_ic_t2m_mou_7          0.00
loc_og_t2t_mou_7          0.00
arpu_7                    0.00
arpu_8                    0.00
arpu_9                    0.00
onnet_mou_7               0.00
onnet_mou_8               0.00
onnet_mou_9               0.00
offnet_mou_7              0.00
offnet_mou_8              0.00
offnet_mou_9              0.00
roam_ic_mou_7             0.00
roam_ic_mou_8             0

#### June

In [147]:
jun_missing_cols = (telecom_df_missing_columns[telecom_df_missing_columns['Null'] == 0.41].index).to_list()
jun_missing_cols

['loc_ic_mou_6',
 'spl_og_mou_6',
 'loc_og_t2c_mou_6',
 'loc_og_mou_6',
 'std_og_t2t_mou_6',
 'std_og_t2m_mou_6',
 'std_og_t2f_mou_6',
 'std_og_mou_6',
 'isd_og_mou_6',
 'og_others_6',
 'loc_og_t2m_mou_6',
 'ic_others_6',
 'loc_ic_t2t_mou_6',
 'isd_ic_mou_6',
 'loc_ic_t2m_mou_6',
 'spl_ic_mou_6',
 'loc_ic_t2f_mou_6',
 'std_ic_t2t_mou_6',
 'std_ic_mou_6',
 'std_ic_t2m_mou_6',
 'loc_og_t2f_mou_6',
 'std_ic_t2f_mou_6',
 'roam_og_mou_6',
 'onnet_mou_6',
 'roam_ic_mou_6',
 'offnet_mou_6',
 'loc_og_t2t_mou_6']

In [148]:
# Creating a dataframe for MOU for August, for which the the MOU values are null
telecom_df_null_mou_6 = telecom_df[(telecom_df['loc_ic_mou_6'].isnull()) & 
                                    (telecom_df['spl_og_mou_6'].isnull()) & 
                                    (telecom_df['loc_og_t2c_mou_6'].isnull()) & 
                                    (telecom_df['loc_og_mou_6'].isnull()) & 
                                    (telecom_df['std_og_t2t_mou_6'].isnull()) & 
                                    (telecom_df['std_og_t2m_mou_6'].isnull()) & 
                                    (telecom_df['std_og_t2f_mou_6'].isnull()) & 
                                    (telecom_df['std_og_mou_6'].isnull()) & 
                                    (telecom_df['isd_og_mou_6'].isnull()) & 
                                    (telecom_df['og_others_6'].isnull()) & 
                                    (telecom_df['loc_og_t2m_mou_6'].isnull()) & 
                                    (telecom_df['ic_others_6'].isnull()) & 
                                    (telecom_df['loc_ic_t2t_mou_6'].isnull()) & 
                                    (telecom_df['isd_ic_mou_6'].isnull()) & 
                                    (telecom_df['loc_ic_t2m_mou_6'].isnull()) & 
                                    (telecom_df['spl_ic_mou_6'].isnull()) & 
                                    (telecom_df['loc_ic_t2f_mou_6'].isnull()) & 
                                    (telecom_df['std_ic_t2t_mou_6'].isnull()) & 
                                    (telecom_df['std_ic_mou_6'].isnull()) & 
                                    (telecom_df['std_ic_t2m_mou_6'].isnull()) &
                                    (telecom_df['loc_og_t2f_mou_6'].isnull()) & 
                                    (telecom_df['std_ic_t2f_mou_6'].isnull()) & 
                                    (telecom_df['roam_og_mou_6'].isnull()) & 
                                    (telecom_df['onnet_mou_6'].isnull()) & 
                                    (telecom_df['roam_ic_mou_6'].isnull()) & 
                                    (telecom_df['offnet_mou_6'].isnull()) & 
                                    (telecom_df['loc_og_t2t_mou_6'].isnull())]

telecom_df_null_mou_6.head()

mobile_number   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
77       7001328263   30.000   82.378  674.950  158.710          NaN   
364      7002168045    0.000  792.112  989.368  923.040          NaN   
423      7000635248  213.802  304.194  149.710  329.643          NaN   
934      7002152278   48.000  764.152  500.030  194.400          NaN   
1187     7000486275    0.000  757.170  995.719    0.000          NaN   

      onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
77          34.23       149.69         6.31           NaN         39.44   
364        433.49       198.96       571.99           NaN        845.11   
423          0.00         0.00         0.00           NaN         10.03   
934         14.24        17.48         7.69           NaN         16.99   
1187      1366.71      2268.91         0.00           NaN          7.78   

      offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
77          179.18         57.68            NaN            0.0           0.00   
364         923.58        828.29            NaN            0.0           0.00   
423           1.45          0.34            NaN            0.0           0.00   
934          76.86         43.64            NaN            0.0           8.81   
1187         36.13          0.00            NaN            0.0           8.08   

      roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
77              0.0            NaN            0.0           0.00   
364             0.0            NaN            0.0           0.00   
423             0.0            NaN            0.0           0.00   
934             0.0            NaN            0.0           1.56   
1187            0.0            NaN            0.0          25.23   

      roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
77             0.00               NaN             34.23            149.69   
364            0.00               NaN             28.78              7.46   
423            0.00               NaN              0.00              0.00   
934            0.00               NaN              0.08             17.48   
1187           0.21               NaN              4.76             46.18   

      loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
77                6.31               NaN             32.18            101.63   
364              64.73               NaN             78.78            584.76   
423               0.00               NaN              0.00              0.58   
934               7.69               NaN             16.99             63.23   
1187              0.00               NaN              7.78             31.29   

      loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
77               29.41               NaN              0.91             29.86   
364             490.71               NaN             21.58              9.43   
423               0.33               NaN              0.00              0.00   
934              39.99               NaN              0.00             12.08   
1187              0.00               NaN              0.00              0.00   

      loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
77               28.26               NaN               0.0               3.9   
364               0.00               NaN               0.0               0.0   
423               0.00               NaN               0.0               0.0   
934               3.65               NaN               0.0               0.0   
1187              0.00               NaN               0.0               0.0   

      loc_og_t2c_mou_9  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  \
77                0.00           NaN         67.33        281.19   
364               2.78           NaN        129.14        601.66   
423               0.00           NaN          0.00          0.58   
934               0.00           NaN         17.

In [149]:
telecom_df_null_mou_6.shape

(115, 171)

In [150]:
# Droping the records for which MOU for June are null
telecom_df = telecom_df.drop(telecom_df_null_mou_6.index)

In [151]:
telecom_df.shape

(27991, 171)

In [152]:
# Let's check the percentage of null values again

telecom_df_missing_columns = (round(((telecom_df.isnull().sum()/len(telecom_df.index))*100),2).to_frame('Null')).sort_values('Null', ascending=False)
telecom_df_missing_columns

Null
date_of_last_rech_9       1.02
date_of_last_rech_8       1.01
date_of_last_rech_7       0.20
date_of_last_rech_6       0.10
mobile_number             0.00
ic_others_6               0.00
spl_ic_mou_7              0.00
spl_ic_mou_8              0.00
spl_ic_mou_9              0.00
isd_ic_mou_6              0.00
isd_ic_mou_7              0.00
isd_ic_mou_8              0.00
isd_ic_mou_9              0.00
ic_others_8               0.00
ic_others_7               0.00
total_rech_amt_7          0.00
ic_others_9               0.00
total_rech_num_6          0.00
total_rech_num_7          0.00
total_rech_num_8          0.00
total_rech_num_9          0.00
spl_ic_mou_6              0.00
total_ic_mou_9            0.00
total_ic_mou_8            0.00
std_ic_t2m_mou_9          0.00
loc_ic_mou_9              0.00
std_ic_t2t_mou_6          0.00
std_ic_t2t_mou_7          0.00
std_ic_t2t_mou_8          0.00
std_ic_t2t_mou_9          0.00
std_ic_t2m_mou_6          0.00
std_ic_t2m_mou_7          0.00
std_ic_t2m_mou_8          0.00
std_ic_t2f_mou_6          0.00
total_ic_mou_7            0.00
std_ic_t2f_mou_7          0.00
std_ic_t2f_mou_8          0.00
std_ic_t2f_mou_9          0.00
std_ic_mou_6              0.00
std_ic_mou_7              0.00
std_ic_mou_8              0.00
std_ic_mou_9              0.00
total_ic_mou_6            0.00
total_rech_amt_6          0.00
max_rech_amt_6            0.00
total_rech_amt_8          0.00
total_rech_amt_9          0.00
sachet_2g_6               0.00
sachet_2g_7               0.00
sachet_2g_8               0.00
sachet_2g_9               0.00
monthly_3g_6              0.00
monthly_3g_7              0.00
monthly_3g_8              0.00
monthly_3g_9              0.00
sachet_3g_6               0.00
sachet_3g_7               0.00
sachet_3g_8               0.00
sachet_3g_9               0.00
aon                       0.00
aug_vbc_3g                0.00
jul_vbc_3g                0.00
jun_vbc_3g                0.00
sep_vbc_3g                0.00
monthly_2g_9              0.00
monthly_2g_8              0.00
monthly_2g_7              0.00
last_day_rch_amt_9        0.00
loc_ic_mou_7              0.00
max_rech_amt_7            0.00
max_rech_amt_8            0.00
max_rech_amt_9            0.00
last_day_rch_amt_6        0.00
last_day_rch_amt_7        0.00
last_day_rch_amt_8        0.00
vol_2g_mb_6               0.00
monthly_2g_6              0.00
vol_2g_mb_7               0.00
vol_2g_mb_8               0.00
vol_2g_mb_9               0.00
vol_3g_mb_6               0.00
vol_3g_mb_7               0.00
vol_3g_mb_8               0.00
vol_3g_mb_9               0.00
loc_ic_mou_8              0.00
loc_ic_mou_6              0.00
arpu_6                    0.00
loc_og_t2f_mou_8          0.00
loc_og_t2t_mou_8          0.00
loc_og_t2t_mou_9          0.00
loc_og_t2m_mou_6          0.00
loc_og_t2m_mou_7          0.00
loc_og_t2m_mou_8          0.00
loc_og_t2m_mou_9          0.00
loc_og_t2f_mou_6          0.00
loc_og_t2f_mou_7          0.00
loc_og_t2f_mou_9          0.00
std_og_t2t_mou_7          0.00
loc_og_t2c_mou_6          0.00
loc_og_t2c_mou_7          0.00
loc_og_t2c_mou_8          0.00
loc_og_t2c_mou_9          0.00
loc_og_mou_6              0.00
loc_og_mou_7              0.00
loc_og_mou_8              0.00
loc_og_mou_9              0.00
loc_og_t2t_mou_7          0.00
loc_og_t2t_mou_6          0.00
roam_og_mou_9             0.00
roam_og_mou_8             0.00
arpu_7                    0.00
arpu_8                    0.00
arpu_9                    0.00
onnet_mou_6               0.00
onnet_mou_7               0.00
onnet_mou_8               0.00
onnet_mou_9               0.00
offnet_mou_6              0.00
offnet_mou_7              0.00
offnet_mou_8              0.00
offnet_mou_9              0.00
roam_ic_mou_6             0.00
roam_ic_mou_7             0.00
roam_ic_mou_8             0.00
roam_ic_mou_9             0.00
roam_og_mou_6             0.00
roam_og_mou_7             0.00
std_og_t2t_mou_6          0.00
std_og_t2t_mou_8          0

#### Tag Churners

Customers are tagged **churned** based on the fourth month as below:
* Customers that did not make any calls or did not receive/have any incoming calls
* Customers who did not use mobile internet in the churn phase

In [153]:
telecom_df.head()

mobile_number   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
8      7001524846  378.721  492.223  137.362  166.787       413.69   
13     7002191713  492.846  205.671  593.260  322.732       501.76   
16     7000875565  430.975  299.869  187.894  206.490        50.51   
17     7000187447  690.008   18.980   25.499  257.583      1185.91   
21     7002124215  514.453  597.753  637.760  578.596       102.41   

    onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
8        351.03        35.08        33.46         94.66         80.63   
13       108.39       534.24       244.81        413.31        119.28   
16        74.01        70.61        31.34        296.29        229.74   
17         9.28         7.79       558.51         61.64          0.00   
21       132.11        85.14       161.63        757.93        896.68   

    offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
8         136.48        108.71           0.00           0.00           0.00   
13        482.46        214.06          23.53         144.24          72.11   
16        162.76        224.39           0.00           2.83           0.00   
17          5.54         87.89           0.00           4.76           4.81   
21        983.39        869.89           0.00           0.00           0.00   

    roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
8            0.00           0.00           0.00           0.00           0.00   
13         136.78           7.98          35.26           1.44          12.78   
16           0.00           0.00          17.74           0.00           0.00   
17           0.00           0.00           8.46          13.34          17.98   
21           0.00           0.00           0.00           0.00           0.00   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
8             297.13            217.59             12.49             26.13   
13             49.63              6.19             36.01              6.14   
16             42.61             65.16             67.38             26.88   
17             38.99              0.00              0.00             36.41   
21              4.48              6.16             23.34             29.98   

    loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
8              80.96             70.58             50.54             34.58   
13            151.13             47.28            294.46            108.24   
16            273.29            145.99            128.28            201.49   
17             58.54              0.00              0.00              9.38   
21             91.81             87.93            104.81            107.54   

    loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
8               0.00              0.00              0.00              0.00   
13              4.54              0.00             23.51              5.29   
16              0.00              4.48             10.26              4.66   
17              0.00              0.00              0.00              0.00   
21              0.75              0.00              1.58              0.00   

    loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  \
8                0.0               0.0              7.15               0.0   
13               0.0               0.0              0.49               0.0   
16               0.0               0.0              0.00               0.0   
17               0.0               0.0              0.00               0.0   
21               0.0               0.0              0.00               0.0   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
8         378.09        288.18         63.04         60.71            116.56   
13        205.31         53.48        353.99        119.69            446.41   
16        315.91        215.64        205.93        233.04        

In [154]:
telecom_df['churn'] = np.where((telecom_df['total_ic_mou_9']==0) & (telecom_df['total_og_mou_9']==0) & (telecom_df['vol_2g_mb_9']==0) & (telecom_df['vol_3g_mb_9']==0), 1, 0)

In [155]:
telecom_df.head()

mobile_number   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
8      7001524846  378.721  492.223  137.362  166.787       413.69   
13     7002191713  492.846  205.671  593.260  322.732       501.76   
16     7000875565  430.975  299.869  187.894  206.490        50.51   
17     7000187447  690.008   18.980   25.499  257.583      1185.91   
21     7002124215  514.453  597.753  637.760  578.596       102.41   

    onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
8        351.03        35.08        33.46         94.66         80.63   
13       108.39       534.24       244.81        413.31        119.28   
16        74.01        70.61        31.34        296.29        229.74   
17         9.28         7.79       558.51         61.64          0.00   
21       132.11        85.14       161.63        757.93        896.68   

    offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
8         136.48        108.71           0.00           0.00           0.00   
13        482.46        214.06          23.53         144.24          72.11   
16        162.76        224.39           0.00           2.83           0.00   
17          5.54         87.89           0.00           4.76           4.81   
21        983.39        869.89           0.00           0.00           0.00   

    roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
8            0.00           0.00           0.00           0.00           0.00   
13         136.78           7.98          35.26           1.44          12.78   
16           0.00           0.00          17.74           0.00           0.00   
17           0.00           0.00           8.46          13.34          17.98   
21           0.00           0.00           0.00           0.00           0.00   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
8             297.13            217.59             12.49             26.13   
13             49.63              6.19             36.01              6.14   
16             42.61             65.16             67.38             26.88   
17             38.99              0.00              0.00             36.41   
21              4.48              6.16             23.34             29.98   

    loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
8              80.96             70.58             50.54             34.58   
13            151.13             47.28            294.46            108.24   
16            273.29            145.99            128.28            201.49   
17             58.54              0.00              0.00              9.38   
21             91.81             87.93            104.81            107.54   

    loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
8               0.00              0.00              0.00              0.00   
13              4.54              0.00             23.51              5.29   
16              0.00              4.48             10.26              4.66   
17              0.00              0.00              0.00              0.00   
21              0.75              0.00              1.58              0.00   

    loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  \
8                0.0               0.0              7.15               0.0   
13               0.0               0.0              0.49               0.0   
16               0.0               0.0              0.00               0.0   
17               0.0               0.0              0.00               0.0   
21               0.0               0.0              0.00               0.0   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
8         378.09        288.18         63.04         60.71            116.56   
13        205.31         53.48        353.99        119.69            446.41   
16        315.91        215.64        205.93        233.04        

#### Delete all the attributes corresponding to the churn phase

In [164]:
# List the columns for churn month: September
column_9 = [col for col in telecom_df.columns.to_list() if '_9' in col]
column_9

['arpu_9',
 'onnet_mou_9',
 'offnet_mou_9',
 'roam_ic_mou_9',
 'roam_og_mou_9',
 'loc_og_t2t_mou_9',
 'loc_og_t2m_mou_9',
 'loc_og_t2f_mou_9',
 'loc_og_t2c_mou_9',
 'loc_og_mou_9',
 'std_og_t2t_mou_9',
 'std_og_t2m_mou_9',
 'std_og_t2f_mou_9',
 'std_og_mou_9',
 'isd_og_mou_9',
 'spl_og_mou_9',
 'og_others_9',
 'total_og_mou_9',
 'loc_ic_t2t_mou_9',
 'loc_ic_t2m_mou_9',
 'loc_ic_t2f_mou_9',
 'loc_ic_mou_9',
 'std_ic_t2t_mou_9',
 'std_ic_t2m_mou_9',
 'std_ic_t2f_mou_9',
 'std_ic_mou_9',
 'total_ic_mou_9',
 'spl_ic_mou_9',
 'isd_ic_mou_9',
 'ic_others_9',
 'total_rech_num_9',
 'total_rech_amt_9',
 'max_rech_amt_9',
 'date_of_last_rech_9',
 'last_day_rch_amt_9',
 'vol_2g_mb_9',
 'vol_3g_mb_9',
 'monthly_2g_9',
 'sachet_2g_9',
 'monthly_3g_9',
 'sachet_3g_9']

In [165]:
# Deleting the september month columns
telecom_df = telecom_df.drop(column_9, axis=1)

In [168]:
telecom_df.head()

mobile_number   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  \
8      7001524846  378.721  492.223  137.362       413.69       351.03   
13     7002191713  492.846  205.671  593.260       501.76       108.39   
16     7000875565  430.975  299.869  187.894        50.51        74.01   
17     7000187447  690.008   18.980   25.499      1185.91         9.28   
21     7002124215  514.453  597.753  637.760       102.41       132.11   

    onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
8         35.08         94.66         80.63        136.48           0.00   
13       534.24        413.31        119.28        482.46          23.53   
16        70.61        296.29        229.74        162.76           0.00   
17         7.79         61.64          0.00          5.54           0.00   
21        85.14        757.93        896.68        983.39           0.00   

    roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
8            0.00           0.00           0.00           0.00           0.00   
13         144.24          72.11           7.98          35.26           1.44   
16           2.83           0.00           0.00          17.74           0.00   
17           4.76           4.81           0.00           8.46          13.34   
21           0.00           0.00           0.00           0.00           0.00   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  \
8             297.13            217.59             12.49             80.96   
13             49.63              6.19             36.01            151.13   
16             42.61             65.16             67.38            273.29   
17             38.99              0.00              0.00             58.54   
21              4.48              6.16             23.34             91.81   

    loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
8              70.58             50.54              0.00              0.00   
13             47.28            294.46              4.54              0.00   
16            145.99            128.28              0.00              4.48   
17              0.00              0.00              0.00              0.00   
21             87.93            104.81              0.75              0.00   

    loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
8               0.00               0.0               0.0              7.15   
13             23.51               0.0               0.0              0.49   
16             10.26               0.0               0.0              0.00   
17              0.00               0.0               0.0              0.00   
21              1.58               0.0               0.0              0.00   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  \
8         378.09        288.18         63.04            116.56   
13        205.31         53.48        353.99            446.41   
16        315.91        215.64        205.93              7.89   
17         97.54          0.00          0.00           1146.91   
21         97.04         94.09        129.74             97.93   

    std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  std_og_t2m_mou_7  \
8             133.43             22.58             13.69             10.04   
13             85.98            498.23            255.36             52.94   
16              2.58              3.23             22.99             64.51   
17              0.81              0.00              1.55              0.00   
21            125.94             61.79            665.36            808.74   

    std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
8              75.69               0.0               0.0               0.0   
13            156.94               0.0               0.0               0.0   
16             18.29               0.0               0.0               0.0   
17              0.00          

In [169]:
# Deleting the sep_vbc_3g column
telecom_df = telecom_df.drop('sep_vbc_3g', axis=1)

In [170]:
telecom_df.shape

(27991, 130)

#### Checking churn percentage

In [161]:
round(100*telecom_df['churn'].mean(),2)

3.39

We can see that the churn percentage is very low at 3.39. This shows that we have an **imbalanced class**. This needs to be handled.

### Feature Engineering

#### Derive New Features

In [172]:
telecom_df.head()

mobile_number   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  \
8      7001524846  378.721  492.223  137.362       413.69       351.03   
13     7002191713  492.846  205.671  593.260       501.76       108.39   
16     7000875565  430.975  299.869  187.894        50.51        74.01   
17     7000187447  690.008   18.980   25.499      1185.91         9.28   
21     7002124215  514.453  597.753  637.760       102.41       132.11   

    onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
8         35.08         94.66         80.63        136.48           0.00   
13       534.24        413.31        119.28        482.46          23.53   
16        70.61        296.29        229.74        162.76           0.00   
17         7.79         61.64          0.00          5.54           0.00   
21        85.14        757.93        896.68        983.39           0.00   

    roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
8            0.00           0.00           0.00           0.00           0.00   
13         144.24          72.11           7.98          35.26           1.44   
16           2.83           0.00           0.00          17.74           0.00   
17           4.76           4.81           0.00           8.46          13.34   
21           0.00           0.00           0.00           0.00           0.00   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  \
8             297.13            217.59             12.49             80.96   
13             49.63              6.19             36.01            151.13   
16             42.61             65.16             67.38            273.29   
17             38.99              0.00              0.00             58.54   
21              4.48              6.16             23.34             91.81   

    loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
8              70.58             50.54              0.00              0.00   
13             47.28            294.46              4.54              0.00   
16            145.99            128.28              0.00              4.48   
17              0.00              0.00              0.00              0.00   
21             87.93            104.81              0.75              0.00   

    loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
8               0.00               0.0               0.0              7.15   
13             23.51               0.0               0.0              0.49   
16             10.26               0.0               0.0              0.00   
17              0.00               0.0               0.0              0.00   
21              1.58               0.0               0.0              0.00   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  \
8         378.09        288.18         63.04            116.56   
13        205.31         53.48        353.99            446.41   
16        315.91        215.64        205.93              7.89   
17         97.54          0.00          0.00           1146.91   
21         97.04         94.09        129.74             97.93   

    std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  std_og_t2m_mou_7  \
8             133.43             22.58             13.69             10.04   
13             85.98            498.23            255.36             52.94   
16              2.58              3.23             22.99             64.51   
17              0.81              0.00              1.55              0.00   
21            125.94             61.79            665.36            808.74   

    std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
8              75.69               0.0               0.0               0.0   
13            156.94               0.0               0.0               0.0   
16             18.29               0.0               0.0               0.0   
17              0.00          

In [198]:
col = [col for col in telecom_df.columns if '_6' not in col and '_7' not in col and '_8' not in col]
col

['mobile_number', 'aon', 'aug_vbc_3g', 'jul_vbc_3g', 'jun_vbc_3g', 'churn']

**Columns for 6th Month:**
['arpu_6',
 'onnet_mou_6',
 'offnet_mou_6',
 'roam_ic_mou_6',
 'roam_og_mou_6',
 'loc_og_t2t_mou_6',
 'loc_og_t2m_mou_6',
 'loc_og_t2f_mou_6',
 'loc_og_t2c_mou_6',
 'loc_og_mou_6',
 'std_og_t2t_mou_6',
 'std_og_t2m_mou_6',
 'std_og_t2f_mou_6',
 'std_og_mou_6',
 'isd_og_mou_6',
 'spl_og_mou_6',
 'og_others_6',
 'total_og_mou_6',
 'loc_ic_t2t_mou_6',
 'loc_ic_t2m_mou_6',
 'loc_ic_t2f_mou_6',
 'loc_ic_mou_6',
 'std_ic_t2t_mou_6',
 'std_ic_t2m_mou_6',
 'std_ic_t2f_mou_6',
 'std_ic_mou_6',
 'total_ic_mou_6',
 'spl_ic_mou_6',
 'isd_ic_mou_6',
 'ic_others_6',
 'total_rech_num_6',
 'total_rech_amt_6',
 'max_rech_amt_6',
 'date_of_last_rech_6',
 'last_day_rch_amt_6',
 'vol_2g_mb_6',
 'vol_3g_mb_6',
 'monthly_2g_6',
 'sachet_2g_6',
 'monthly_3g_6',
 'sachet_3g_6',
 'Average_rech_amt_6_and_7']


**Columns for 7th Month:**
['arpu_7',
 'onnet_mou_7',
 'offnet_mou_7',
 'roam_ic_mou_7',
 'roam_og_mou_7',
 'loc_og_t2t_mou_7',
 'loc_og_t2m_mou_7',
 'loc_og_t2f_mou_7',
 'loc_og_t2c_mou_7',
 'loc_og_mou_7',
 'std_og_t2t_mou_7',
 'std_og_t2m_mou_7',
 'std_og_t2f_mou_7',
 'std_og_mou_7',
 'isd_og_mou_7',
 'spl_og_mou_7',
 'og_others_7',
 'total_og_mou_7',
 'loc_ic_t2t_mou_7',
 'loc_ic_t2m_mou_7',
 'loc_ic_t2f_mou_7',
 'loc_ic_mou_7',
 'std_ic_t2t_mou_7',
 'std_ic_t2m_mou_7',
 'std_ic_t2f_mou_7',
 'std_ic_mou_7',
 'total_ic_mou_7',
 'spl_ic_mou_7',
 'isd_ic_mou_7',
 'ic_others_7',
 'total_rech_num_7',
 'total_rech_amt_7',
 'max_rech_amt_7',
 'date_of_last_rech_7',
 'last_day_rch_amt_7',
 'vol_2g_mb_7',
 'vol_3g_mb_7',
 'monthly_2g_7',
 'sachet_2g_7',
 'monthly_3g_7',
 'sachet_3g_7',
 'Average_rech_amt_6_and_7']

**Columns for 8th Month:**

['arpu_8',
 'onnet_mou_8',
 'offnet_mou_8',
 'roam_ic_mou_8',
 'roam_og_mou_8',
 'loc_og_t2t_mou_8',
 'loc_og_t2m_mou_8',
 'loc_og_t2f_mou_8',
 'loc_og_t2c_mou_8',
 'loc_og_mou_8',
 'std_og_t2t_mou_8',
 'std_og_t2m_mou_8',
 'std_og_t2f_mou_8',
 'std_og_mou_8',
 'isd_og_mou_8',
 'spl_og_mou_8',
 'og_others_8',
 'total_og_mou_8',
 'loc_ic_t2t_mou_8',
 'loc_ic_t2m_mou_8',
 'loc_ic_t2f_mou_8',
 'loc_ic_mou_8',
 'std_ic_t2t_mou_8',
 'std_ic_t2m_mou_8',
 'std_ic_t2f_mou_8',
 'std_ic_mou_8',
 'total_ic_mou_8',
 'spl_ic_mou_8',
 'isd_ic_mou_8',
 'ic_others_8',
 'total_rech_num_8',
 'total_rech_amt_8',
 'max_rech_amt_8',
 'date_of_last_rech_8',
 'last_day_rch_amt_8',
 'vol_2g_mb_8',
 'vol_3g_mb_8',
 'monthly_2g_8',
 'sachet_2g_8',
 'monthly_3g_8',
 'sachet_3g_8']

**Other non month columns**
['mobile_number', 'aon', 'aug_vbc_3g', 'jul_vbc_3g', 'jun_vbc_3g', 'churn']

** Details about the columns to help in deriving new features **
* arpu_6: Average revenue per user 
* onnet_mou_6: All kind of calls within the same operator network 
* offnet_mou_6: All kind of calls outside the operator T network 
* roam_ic_mou_6: Indicates that customer is in roaming zone during the call 
* roam_og_mou_6: Indicates that customer is in roaming zone during the call 
* loc_og_t2t_mou_6: Local, outgoing calls withing same operator MOU 
* loc_og_t2m_mou_6: Local, outgoing calls to other mobile MOU 
* loc_og_t2f_mou_6: Local, outgoing calls to fixed lines MOU 
* loc_og_t2c_mou_6: Local, outgoing calls to own call center MOU 
* loc_og_mou_6: Local, outgoing calls MOU 
* std_og_t2t_mou_6: STD, outgoing call withing same operator MOU 
* std_og_t2m_mou_6: STD, outgoing call to other mobile MOU 
* std_og_t2f_mou_6: STD, outgoing call to fixed lines MOU 
* std_og_mou_6: STD, outgoing call MOU 
* isd_og_mou_6: ISD outgoing call, MOU 
* spl_og_mou_6: Special outgoing call, MOU 
* og_others_6: Outgoing call, others MOU 
* total_og_mou_6: Total outgoing call, MOU 
* loc_ic_t2t_mou_6: Local, incoming calls withing same operator MOU 
* loc_ic_t2m_mou_6: Local, incoming calls to other mobile MOU 
* loc_ic_t2f_mou_6: Local, incoming calls to fixed lines MOU 
* loc_ic_mou_6: Local, incoming calls MOU 
* std_ic_t2t_mou_6: STD, incoming call withing same operator MOU 
* std_ic_t2m_mou_6: STD, incoming call to other mobile MOU 
* std_ic_t2f_mou_6: STD, incoming call to fixed lines MOU 
* std_ic_mou_6: STD, incoming MOU 
* total_ic_mou_6: Total incoming call, MOU 
* spl_ic_mou_6: Special incoming call, MOU 
* isd_ic_mou_6: ISD Incoming call MOU 
* ic_others_6: Incoming call, others MOU
* total_rech_num_6: Total Recharge Number
* total_rech_amt_6: Total Recharge Amount 
* max_rech_amt_6: Maximum Recharge Amount 
* date_of_last_rech_6: Date of last recharge 
* last_day_rch_amt_6: Last Day Recharge Amount 
* vol_2g_mb_6: Mobile internet usage volumn (2g) 
* vol_3g_mb_6: Mobile internet usage volumn (3g) 
* monthly_2g_6: Service schemes with validity equivalent to a month (2g) 
* sachet_2g_6: Service schemes with validity smaller than a month (2g)
* monthly_3g_6: Service schemes with validity equivalent to a month (2g)
* sachet_3g_6: Service schemes with validity smaller than a month (3g)
* Average_rech_amt_6_and_7

**Some of the derived features that can be created are:**
* Reduction in recharge amount
* Reduction in recharge number
* Reduction in mobile internet usage
* Reduction in average revenue per user
* Reduction in the MOU of the customer

#### Derive new column to check reduction/increment in recharge amount

In [201]:
# Average recharge_amount in good phase
telecom_df['avg_rech_amt_good'] = (telecom_df['total_rech_amt_6'] + telecom_df['total_rech_amt_7'])/2

In [204]:
# Difference of good phase recharge amtount and action phase recharge amtount
telecom_df['diff_rech_amt'] = telecom_df['avg_rech_amt_good'] - telecom_df['total_rech_amt_8']

In [207]:
# Checking if recharge_amount has decreased in action phase
telecom_df['decrease_rech_amt_action'] = np.where((telecom_df['diff_rech_amt'] < 0), 1, 0) 

In [208]:
telecom_df.head()

mobile_number   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  \
8      7001524846  378.721  492.223  137.362       413.69       351.03   
13     7002191713  492.846  205.671  593.260       501.76       108.39   
16     7000875565  430.975  299.869  187.894        50.51        74.01   
17     7000187447  690.008   18.980   25.499      1185.91         9.28   
21     7002124215  514.453  597.753  637.760       102.41       132.11   

    onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
8         35.08         94.66         80.63        136.48           0.00   
13       534.24        413.31        119.28        482.46          23.53   
16        70.61        296.29        229.74        162.76           0.00   
17         7.79         61.64          0.00          5.54           0.00   
21        85.14        757.93        896.68        983.39           0.00   

    roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
8            0.00           0.00           0.00           0.00           0.00   
13         144.24          72.11           7.98          35.26           1.44   
16           2.83           0.00           0.00          17.74           0.00   
17           4.76           4.81           0.00           8.46          13.34   
21           0.00           0.00           0.00           0.00           0.00   

    loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  \
8             297.13            217.59             12.49             80.96   
13             49.63              6.19             36.01            151.13   
16             42.61             65.16             67.38            273.29   
17             38.99              0.00              0.00             58.54   
21              4.48              6.16             23.34             91.81   

    loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
8              70.58             50.54              0.00              0.00   
13             47.28            294.46              4.54              0.00   
16            145.99            128.28              0.00              4.48   
17              0.00              0.00              0.00              0.00   
21             87.93            104.81              0.75              0.00   

    loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
8               0.00               0.0               0.0              7.15   
13             23.51               0.0               0.0              0.49   
16             10.26               0.0               0.0              0.00   
17              0.00               0.0               0.0              0.00   
21              1.58               0.0               0.0              0.00   

    loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  \
8         378.09        288.18         63.04            116.56   
13        205.31         53.48        353.99            446.41   
16        315.91        215.64        205.93              7.89   
17         97.54          0.00          0.00           1146.91   
21         97.04         94.09        129.74             97.93   

    std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  std_og_t2m_mou_7  \
8             133.43             22.58             13.69             10.04   
13             85.98            498.23            255.36             52.94   
16              2.58              3.23             22.99             64.51   
17              0.81              0.00              1.55              0.00   
21            125.94             61.79            665.36            808.74   

    std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
8              75.69               0.0               0.0               0.0   
13            156.94               0.0               0.0               0.0   
16             18.29               0.0               0.0               0.0   
17              0.00          

In [209]:
telecom_df['decrease_rech_amt_action'].mean()

0.40384409274409633